Input layer of a fixed shape for VMModel

vhavrylov-ncube opened the pull request:
- Model can now have an Input layer of a given (fixed) shape
- all other shapes for Dense layers and Ops-layers (Flatten, Dense, ...) are deduced from the single InputLayer shape
- ml node functionality partially transferred to .cpp file from header-only.
- Layer (Nodes) input and output shapes are automatically inferred/deduced using the Input layer shape
- refactored graph compilatioon process
- refactored Dense layer
- added tests

vhavrylov-ncube committed:
InputLayer skeleton added

vhavrylov-ncube committed:
Unit test fixed

vhavrylov-ncube committed:
Initial implementation of shape linking from Input

vhavrylov-ncube committed:
Shape deduction skeleton

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
CLI header corrected

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Auto deduction of input count for Dense skeleton

vhavrylov-ncube committed:
Dense shape auto-detection polishing

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Adapted to reworked Weights op

vhavrylov-ncube committed:
Code cleanup

baykaner commented on libs/ml/include/ml/layers/fully_connected.hpp:
```diff
@@ -75,43 +75,65 @@ class FullyConnected : public SubGraph<T>
+
+    // When expected input shape is known, it is possible to compute flattened input
+    // shape of this fully connected layer
+    this->nodes_.at(flat_input_)
```
flat_input_ is a confusing naming choice given how it's now being used

vhavrylov-ncube replied:
input_ and flat_input_ could be the same if the real input Tensor is 2-dimensional (e.g. data is plain 1 dim vector) and could be 2 different layers if this Dense is time_distributed (I don't know what does it mean, but the code says so :) 
I think we could name this node "flattened_input_" or alike.

baykaner commented on libs/ml/include/ml/model/sequential.hpp:
```diff
@@ -47,6 +47,8 @@ class Sequential : public Model<TensorType>
 
   SizeType LayerCount() const;
 
+  void SetExpectedInputShape(std::vector<math::SizeType> const &shape);
```
naming: drop the 'expected'.

vhavrylov-ncube replied:
done

baykaner commented on libs/ml/include/ml/ops/ops.hpp:
```diff
@@ -73,8 +95,128 @@ class Ops
     return is_training_;
   }
 
+  virtual void SetSliceOutputShape(Shape const &new_shape)
```
I don't like all of these XSliceX names. I understand why you're saying slice, but in general a slice can be across any dimension; you're talking about a very particular way of slicing - batch slicing. Maybe these should all be called SetBatchOutputShape, ComputeBatchOutputShape etc.?

vhavrylov-ncube replied:
Batch is a nice name too!

baykaner commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,147 @@
+ * @return vector of SizeType.
+ */
+template <typename TensorType>
+Shape Node<TensorType>::SliceOutputShape()
```
BatchOutputShape

vhavrylov-ncube replied:
done

baykaner commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,147 @@
+  if (input_nodes_.empty())
+  {
+    FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a Graph leaf : " + this->name_);
+    return candidate;
```
I suspect static analysis won't let you do this pattern anyway:
if (X)
{
  return ...
}

if (Y)
{
  return ...
}

in any case it would be clearer if there were else statements

vhavrylov-ncube replied:
I'll double check static analysis behavior in this case.

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
shape members & methods renaming

vhavrylov-ncube committed:
Renaming, some tests fixed, WIP

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
More tests fixed, WIP

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
more tests fixed, added OOP wrappers

vhavrylov-ncube committed:
UT for VMModel corrected

vhavrylov-ncube committed:
Tests fixed: CEL and SCEL updated

vhavrylov-ncube committed:
More comments and code cleanup

vhavrylov-ncube committed:
WIP: shape autodetect adhoc patches

vhavrylov-ncube committed:
Merge branch 'master' of https://github.com/fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Separated failing UTs for time-distributed FC

+ added more OpType/Descriptor getters
+ all other UTs pass ok

vhavrylov-ncube committed:
Restored Dense input count assertion

vhavrylov-ncube committed:
Disabled a deprecated Classifier unit test

vhavrylov-ncube committed:
Preliminary fix for time-distributed

+ fixed time-distributed Denses crash in some tests
+ added more OOP wrappers for opcode and descriptor in Ops

vhavrylov-ncube committed:
Reverted unnecessary change

vhavrylov-ncube committed:
Comments updated

vhavrylov-ncube committed:
OpType retrieving refactored, ShapeAsString moved

+ added dynamic-pointer-cast-based check for is the Op a Subgraph
+ moved "AsString" methods from Ops to NOde (are not used in Ops)
+ added more comments
+ added serialisation of Shapes inside the Ops

vhavrylov-ncube committed:
WIP: fixed SkipGram op rebuild test BREAKS OTHER TESTS!

vhavrylov-ncube committed:
Fixed re-broken tests

vhavrylov-ncube committed:
Fixed all failing tests, added test for Dense autoinputs

vhavrylov-ncube commented on libs/ml/include/ml/core/graph.hpp:
```diff
@@ -811,7 +844,7 @@ void Graph<TensorType>::SetInputReference(std::string const &node_name, TensorTy
 void Graph<TensorType>::SetInput(std::string const &node_name, TensorType const &data)
 {
-  SetInputReference(node_name, data.Copy());
+  SetInputReference(node_name, data);
```
copying is redundant here, because SetInputReference takes its argument by const ref, and "data" is already a constant ref here.

vhavrylov-ncube committed:
Added mixed layers auto-shaping test

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

iryna-melnyk commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,223 @@
+  {
+    FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a Graph leaf : " + this->name_);
+    return candidate;
+  }
```
I would suggest to rearrange branching order and put check for emptiness first (because you actually called it more than once which is not necessary IMHO) it may improve the readability

vhavrylov-ncube replied:
done!

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Refactored Node shaping method for better readability

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Removed obsolete test

baykaner commented on libs/ml/include/ml/core/node.hpp:
```diff
@@ -397,14 +396,17 @@ void Node<TensorType>::ResetCache(bool input_size_changed)
   op_ptr_               = op_ptr;
 }
 
+std::string OutputShapeAsString(fetch::math::SizeVector const &out_shape);
```
these two functions look like utilities.
Firstly, I believe they should be static functions.
Secondly, they should either live only in the cpp (if we're sure they will never be invoked from anywhere except node), or they should move to a more general utilities location (like ml/utilities).

baykaner commented on libs/ml/include/ml/layers/PRelu.hpp:
```diff
@@ -99,6 +99,15 @@ class PRelu : public SubGraph<T>
   }
 
   static constexpr char const *DESCRIPTOR = "ParametricRelu";
```
presumably we can ditch this and just have the Descriptor function?

vhavrylov-ncube replied:
I believe yes, but for now ::DESCRIPTOR is used in a template code 

template \<typename OperationType\>
bool Graph\<TensorType\>::UpdateVariableName(
...
  std::string op_descriptor = (OperationType::DESCRIPTOR);

so I guess we will ditch both DESCRIPTOR and OpType fields and replace them by virtual polymorphic functions after refactoring of GraphBuilder and that code above.

baykaner commented on libs/ml/include/ml/layers/fully_connected.hpp:
```diff
@@ -189,22 +259,66 @@ class FullyConnected : public SubGraph<T>
-    fetch::ml::ops::Weights<TensorType>::Initialise(weights, in_size_, out_size_, init_mode);
-  }
+  // Non-saveable params
+  bool                          is_initialised_ = false;
```
all these non-saveable params leave me slightly nervous about whether dense layer serialisation->deserialisation will ALWAYS result in the same behaviour.

For example. is_initialised_ can be true for a dense layer. Serialise and deserialise the layer and the new dense layer has is_initialised_ = false. Couldn't that be a cause of differing behaviour in the layers?

baykaner replied:
the safe (but also less efficient) thing to do is serialised all the parameters - then behaviour is guaranteed; but you're right that there are definitely some parameters that are just effectively internal caching and don't need to be serialised.

vhavrylov-ncube replied:
I agree. I have removed all fields possible, leaving only what are really necessary to keep, and serialised them.

baykaner commented on libs/ml/include/ml/ops/ops.hpp:
```diff
@@ -52,14 +54,34 @@ class Ops
+   */
+  virtual std::vector<SizeType> ComputeBatchOutputShape(ShapeVector const &input_shapes)
+  {
+    VecTensorType dummies;
```
inefficient formulation. More efficient formulation is either to construct using input_shape.size() and then emplace back OR use vector.reserve() and then assign dummies[i] = ...
emplace back is probably neater

vhavrylov-ncube replied:
done

baykaner commented on libs/ml/include/ml/saveparams/saveable_params.hpp:
```diff
@@ -37,8 +37,10 @@ struct OpsSaveableParams
-  bool              is_training = true;
+  fetch::ml::OpType                    op_type     = OpType::NONE;
+  bool                                 is_training = true;
+  std::vector<fetch::math::SizeVector> batch_input_shapes{};
```
if you're adding these two parameters to saveable params then you need to update the serialisers affected by this right?

vhavrylov-ncube replied:
oh, I've missed it, sure!

baykaner commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,236 @@
+Shape Node<TensorType>::BatchOutputShape()
+{
+  Shape const &candidate = op_ptr_->BatchOutputShape();
+  bool const   op_is_subgraph =
```
this is an unnecessary step. you could just:

auto subgraph_ptr = dynamic_pointer_cast<Subgraph>(op_ptr_);

if (!candidate.empty() && subgraph_ptr)
{

}

baykaner replied:
if it's a matter of clarity I would say that the comment on line 84 handles that for you - "if the op is not a subgraph ..."

vhavrylov-ncube replied:
agreed

baykaner commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,236 @@
+      continue;
+    }
+
+    input_shapes.emplace_back(in_shape);
```
i think it's a little clearer to formulate it like this:

if (!in_shape.empty())
{
  input_shapes.emplace_back(in_shape);
}
else
{
  // error handling...
}

vhavrylov-ncube replied:
In this method I've tried to keep the next approach: if something went wrong, we return, else we continue. This way each next instruction is executed only if there were no errors before. Keeping an "else" block will result in less clearer flow here, as it is a bit harder to understand which of the blocks ("if" or "else" one) is for good case.

baykaner commented on libs/ml/src/core/node.cpp:
```diff
@@ -0,0 +1,236 @@
+template class Node<math::Tensor<uint8_t>>;
+template class Node<math::Tensor<uint16_t>>;
+template class Node<math::Tensor<uint32_t>>;
+template class Node<math::Tensor<uint64_t>>;
```
don't need these

vhavrylov-ncube replied:
removed

baykaner commented on libs/ml/tests/unit/graph/graph_every_op.cpp:
```diff
@@ -114,7 +114,83 @@ void ComparePrediction(GraphPtrType g, GraphPtrType g2, std::string node_name)
 }
 
-TYPED_TEST(GraphRebuildTest, graph_rebuild_every_op)
+TYPED_TEST(GraphRebuildTest, graph_rebuild_time_distributed_ops)
```
why do we separate out these tests? is this to do with time distributed not working with shape deduction currently? will they be merged back together after?

vhavrylov-ncube replied:
well, the reason for separation was - time-distributed ones failed until I fixed them later. Now I just wonder if it would be better to split that "rebuild-all" approach to a smaller groups or even individual tests. But you're right it looks like a separate task, not related to current change, so I'll just merge it back for now.

baykaner commented on libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:
```diff
@@ -1579,4 +1591,88 @@ TEST_F(VMModelTests, model_fit_and_refit)
+
+         var model = Model("sequential");
+
+         model.addExperimental("input", data_shape);
```
so the input layer is working; but we don't yet have the right bindings for dense and conv to allow just output shape specification?
But presumably all the underlying code is there for that, we just need to add a couple of functions to vmmodules sequential, right?

vhavrylov-ncube replied:
model.addExperimental("dense", 6u64);  -- this is already here for output-only Dense declaration; regarding Conv2D - it has only "input channels" parameter, it might be auto-deduced from some input shape of previous layer, but I am not ready to tell if this implementation is trivial.

vhavrylov-ncube replied:
Well, the first leyer in a Model could be also Dropout, Flatten, Reshape or another one, so it moght be a good idea to create interface for adding any type of layer to a Model with ptoviding input shape array as a parameter (as it is possible in Keras), however, it is a large change so I would prefer to keep only Dense as an example and an InputLayer as a default working solution for now.

vhavrylov-ncube committed:
Added -nochargelimit debug parameter to etch app

vhavrylov-ncube committed:
PR corrections

vhavrylov-ncube committed:
Added some tests for shape auto-deduction

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Merging master completed

vhavrylov-ncube committed:
More effective ComputeBatchOutputShape

vhavrylov-ncube committed:
Tiny corrections

vhavrylov-ncube committed:
WIP adding serialisers for shapes in Op

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Serialising of in-output shapes added

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
WIP: fixing failing serialisation tests

vhavrylov-ncube committed:
Fixed failing Conv1d and Conv2d tests

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
WIP shared layers test skeleton corrected

vhavrylov-ncube committed:
refactored FC layer construction

vhavrylov-ncube committed:
Removed some saveable params, refactored FC init

vhavrylov-ncube committed:
Added serialisation, simplified shaping

vhavrylov-ncube committed:
Fixed regularisation Compile call

vhavrylov-ncube committed:
Reverted useless change

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
Lint corrections

vhavrylov-ncube committed:
Throw instead of null return

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

vhavrylov-ncube committed:
tiny renaming

vhavrylov-ncube committed:
Merge branch 'master' of github.com:fetchai/ledger into feature/ML-437-InputLayer-fixedshape

baykaner committed:
fixed fully connected layer shared layer not sharing connetions

vhavrylov-ncube reviewed (commented):
approved from my side

baykaner committed:
removed erroneously added includes

fetch-bot committed:
Merge branch 'master' into feature/ML-437-InputLayer-fixedshape
